xds/src/main/java/io/grpc/xds/LoadStatsManager2.java

```diff
@@ -353,6 +357,18 @@ void recordCallFinished(Status status) {
       }
     }
 
+    /**
+     * Records a custom named backend load metric stat for per-call load reporting.
+     */
+    void recordBackendLoadMetricStats(String name, double value) {
```

Comment:
To me the code would look much simpler if you just have an AtomicLong and an AtomicDouble for these two metrics stats, then only create `BackendLoadMetricStats` when snapshot().

Reply:
### Context
Perhaps my knowledge is lacking or I am misunderstanding your comment. Starting with some context, we need to maintain some sort of mapping as there can be different Long & Double values **_for each_** String name. 

For example, a few approaches: concurrent map of `String -> SomeDataStruct<Long, Double>` or 2 concurrent maps of `String -> AtomicLong, String -> AtomicDouble`.

I think the root issue/requirement for `recordBackendLoadMetricStats(String name, double value)` is to _**atomically in one operation handle both**_ :
1. update/increment existing entry with the Long & Double values if the String name exists in the map(s)
2. create a new entry with the Long & Double values if the String name is not found in the map(s). 

Hence, I chose the `merge` operation in the original approach as it can handle all requirements above atomically in one operation to maintain thread safety.

---

### I experimented in code regarding your comment, if I am understanding correctly, would you be proposing we have two concurrent maps instead (e.g. map1 for `String -> AtomicLong` and map2 for `String -> AtomicDouble`)? 

We still need to be careful of atomic updates & thread safety with this new approach. So for `recordBackendLoadMetricStats(String name, double value)`, it would be something along these lines:
```
map1.get().computeIfAbsent(name, k -> new AtomicLong(0L)).incrementAndGet();
map2.get().computeIfAbsent(name, k -> new AtomicDouble(0D)).addAndGet(value);
```
Then, we only create the `BackendLoadMetricStats` in the `snapshot()` method based on the two maps.

---

I agree that with this new approach,  this part of the code is much cleaner, but I am unsure if it is 100% thread safe & prevent race-conditions.

1. Would the chaining of `computeIfAbsent().incrementAndGet()` be thread safe and not introduce a race condition? I think **_it is thread-safe_** since the value of the map is `AtomicLong` meaning it is `volatile` and should be visible in all threads in the case when there is a context switch between the first and second chain method (similar for map2).

2. Would there be an issue if there is a context switch after map1 is fully updated, but before map2 is updated? I am unsure about this part (as they are two separate operations but each operation is atomic). Would there be a case if the two maps are out-of-sync when `snapshot()` is called?

With the original approach, we can guarantee that everything is in-sync since the update/increment if the key exists & create if key does not exist in the map for both Long and Double is handled in one atomic operation via `merge()`

Reply:
I experimented in code with your suggestion of using `computeIfPresent`, specifically
```
loadMetricStatsMap.get().computeIfPresent(name, (k, v) -> BackendLoadMetricStats.create(v.numRequestsFinishedWithMetric() + 1, v.totalMetricValue() + value));
```
However, this _**does not work**_ (and tests fail) because `computeIfPresent(k, remapFn)` [1, 2, 3] behaves:
* If the key is already present in the map, a new value is computed with the remap function (and updated in the map) and returned
* If the key is not present, then `computeIfPresent` returns `null` (**_nothing is updated in the map_**)

Thus, any new key-value pairs are never added if the key was never in the map, so the map remains empty every time `recordBackendLoadMetricStats(String name, double value)` is called. 

To fix this, we would need multiple operations (although each operation itself is atomic). For example, using if/else block and/or combining `computeIfAbsent` and `computeIfPresent`. Again, this requires extra synchronization via the `synchronized` keyword. I would argue it does _**not increase readability**_ (and probably has a slight performance penalty) compared  to `merge`. In fact, other users encountered similar problems and the recommended answer from the community is to use `merge` instead [4, 5].

Other alternatives include using `compute` but that is even more verbose than `merge`, as one needs to write the update/increment & create logic as part of the mapping function (which `merge` implicitly handles) [5].

After examining all ConcurrentHashMap APIs, i think `merge` would be **_most suitable_** since it is like a “subset” of the `compute` operation especially designed to handle **_both requirements atomically in one operation_**:
1. update/increment existing entry with the Long & Double values if the String name exists in the map
2. create a new entry with the Long & Double values if the String name is not found in the map

---
References
[1]https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html#computeIfPresent(K,java.util.function.BiFunction) 
[2]https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Map.html#computeIfPresent(K,java.util.function.BiFunction) 
[3] https://www.educative.io/answers/what-is-the-hashmap-computeifpresent-method-in-java 
[4] https://stackoverflow.com/questions/66747746/refactoring-if-else-block-to-use-map-computeifabsent-map-computeifpresent-wrong 
[5] https://nurkiewicz.com/2019/03/mapmerge-one-method-to-rule-them-all.html

Reply:
Oh sry my suggestion seems bad. It should be:
```
loadMetricStatsMap.get.compute(name, (k, v) -> v == null? new BackendLoadMetricsStats(1, value) : new BackendLoadMetricStats(v.numRequestsFinished + 1, v.totalMetricValue+value))
```
correct me again if I am wrong.

Reply:
Based on our conversation it would worth a comment: For the named metrics key, increment the counter and adds the value, and do that atomically.

Reply:
We decided offline to keep the `merge` approach (as opposed to `compute`). 

In general, the `compute` operation is more flexible as it can be any user-defined mapping function. 
For our specific use case (handling increment/update or create), `merge` could be more appropriate as for this scenario the official Java documentation also recommends using `merge` over `compute` [1].

Added details in latest commit to the Javadoc for `recordBackendLoadMetricStats` documenting the method is atomically performing increment/update or create in one operation.

References
[1]https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/HashMap.html#compute(K,java.util.function.BiFunction)

Reply:
compute() has the advantage of generating less garbage, although overall this will be producing a lot of garbage anyway. Can we not use a mutable value? C uses a lock here, IIRC.

FYI, Java 8 and 17 don't include that note about merge being simpler. I agree it could be simpler, but avoiding needless garbage is probably more important to us for this code.
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#compute-K-java.util.function.BiFunction-
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html#compute(K,java.util.function.BiFunction)

Reply:
Done.
The following steps were done to meet the requirements of less garbage created:
1. `BackendLoadMetricStats` data structure is mutable.
  Previously was immutable due to using `@AutoValue`, now is a class with a method `addMetricValueAndIncrementRequestsFinished(double metricValue)` to modify existing object.
2. `recordBackendLoadMetricStats(Map<String, Double>)` is a `synchronized` method (equivalent to a `synchronized (this) {...}` block inside the entire method) and a synchronized block (`synchronized (this) {...}`) is added for getting and resetting/clearing the map in the `snapshot()` method for thread safety. 
This is equivalent to C++, where there is a lock that is acquired before iterating the named metrics map in order to create/populate values in the backend_metrics map [1]
3. Since locks are used to ensure thread safety, there is no need for the backend load stats map to use `AtomicReference` and `ConcurrentHashMap` as it would just add extra overhead. Hence, it is changed to use a standard `Map`. 
4. Furthermore, there is no need to use `compute()` or `merge()` in the `recordBackendLoadMetricStats()` method, a standard loop with if/else statements is used for better readability.

References
[1] [google3/third_party/grpc/src/core/ext/xds/xds_client_stats.cc;l=167-169](http://google3/third_party/grpc/src/core/ext/xds/xds_client_stats.cc;l=167-169)